src/compiler/utilities.ts

```diff
@@ -2414,6 +2414,27 @@ namespace ts {
         return output;
     }
 
+    export const stringify: (value: any) => string = JSON && JSON.stringify ? JSON.stringify : function stringify(value: any): string {
+        /* tslint:disable:no-null */
+        return value == null ? "null"

```

Comment:
Should use `===` when comparing with `null` - as written this lumps `undefined` with `null`. If `value` is `undefined`, `JSON.stringify` should return `undefined`. A key with an `undefined` value should be omitted, but an array with `undefined` slots should have those slots serialized as `null` (as this does). Presently, a map with `undefined` valued keys gets encoded with `null` values, since `Object.prototype.hasOwnProperty.call({x: undefined}, 'x')` returns `true`.